state/protocol/kvstore.go

```diff
@@ -0,0 +1,71 @@
+	// from the encoded bytes.
+	// No errors are expected during normal operation.
+	VersionedEncode() (uint64, []byte, error)
+}
+
+// ViewBasedActivator allows setting value that will be active from specific view.
+type ViewBasedActivator[T any] struct {
+	Data           T
+	ActivationView uint64 // first view at which Data will take effect
+}
+
+// This file contains versioned read and read-write interfaces to the Protocol State's
+// key-value store and are used by the Protocol State Machine.
+//
+// When a key is added or removed, this requires a new protocol state version:
+//  - Create a new versioned model in ./kvstore/models.go (eg. modelv3 if latest model is modelv2)
+//  - Update the KVStoreReader and KVStoreAPI interfaces to include any new keys
+
+// KVStoreReader is the latest read-only interface to the Protocol State key-value store
+// at a particular block.
+//
+// Caution:
+// Engineers evolving this interface must ensure that it is backwards-compatible
+// with all versions of Protocol State Snapshots that can be retrieved from the local
+// database, which should exactly correspond to the versioned model types defined in
+// ./kvstore/models.go
+type KVStoreReader interface {
+	// ID returns an identifier for this key-value store snapshot by hashing internal fields.
+	ID() flow.Identifier
+
+	// v0
+
+	VersionedEncodable
+
+	// GetProtocolStateVersion returns the Protocol State Version that created the specific
+	// Snapshot backing this interface instance. Slightly simplified, the Protocol State
+	// Version defines the key-value store's data model (specifically, the set of all keys
+	// and the respective type for each corresponding value).
+	// Generally, changes in the protocol state version correspond to changes in the set
+	// of key-value pairs which are supported, and which model is used for serialization.
+	// The protocol state version is updated by UpdateKVStoreVersion service events.
+	GetProtocolStateVersion() uint64
+
+	// GetVersionUpgrade returns the upgrade version of protocol.
+	// VersionUpgrade is a view-based activator that specifies the version which has to be applied
+	// and the view from which on it has to be applied. It may return the current protocol version
+	// with a past view if the upgrade has already been activated.
+	GetVersionUpgrade() *ViewBasedActivator[uint64]
+
+	// GetEpochStateID returns the state ID of the epoch state.
+	// This is part of the most basic model and is used to commit the epoch state to the KV store.
+	GetEpochStateID() flow.Identifier
+
+	// v1
+
+	// GetInvalidEpochTransitionAttempted returns a flag indicating whether epoch
+	// fallback mode has been tentatively triggered on this fork.
+	// Errors:
+	//  - ErrKeyNotSupported if the respective entry does not exist in the
+	//    Protocol State Snapshot that is backing the `Reader` interface.
+	GetInvalidEpochTransitionAttempted() (bool, error)
```

Comment:
I added this originally mainly as a way to have two different versions we could test with right away. But we still have the `InvalidEpochTransitionAttempted` field in the `EpochStateContainer` (where is was originally) and this is what is used for triggering EFM in the state machine. 

I don't think we have an immediate plan to use this version of the flag instead, so I think we should remove it to avoid confusion. 

However, it's still necessary for several tests (in particular the integration test in https://github.com/onflow/flow-go/pull/5477) to have multiple versions to transition between _before releasing the feature_. Not sure what the best course of action is. 

**Suggestion:** At the moment, I'm leaning toward having a v1 model version which is otherwise identical to v0, but returns a different version number.

This comment isn't really related to this PR - feel free to address here, otherwise I can in https://github.com/onflow/flow-go/pull/5477.

Reply:
📝 added item to [issue 5666](https://github.com/onflow/flow-go/issues/5666)